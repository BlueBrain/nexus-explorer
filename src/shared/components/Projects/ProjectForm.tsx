import * as React from 'react';
import { Form, Icon, Input, Button, Spin } from 'antd';
import { WrappedFormUtils } from 'antd/lib/form/Form';

/**
 * Custom from controls, based on:
 * https://ant.design/components/form/#components-form-demo-customized-form-controls
 */
export interface PrefixMappingGroupInputState {
  prefix: string;
  namespace: string;
}
export interface PrefixMappingGroupInputProps {
  groupId: string | number;
  value: any;
  onChange(state: PrefixMappingGroupInputState): void;
}
class PrefixMappingGroupInput extends React.Component<
  PrefixMappingGroupInputProps,
  PrefixMappingGroupInputState
> {
  // Static method called by Form component
  static getDerivedStateFromProps(nextProps: PrefixMappingGroupInputProps) {
    if ('value' in nextProps) {
      return {
        ...(nextProps.value || {}),
      };
    }
    return null;
  }

  constructor(props: PrefixMappingGroupInputProps) {
    super(props);
    const value = props.value || {};
    this.state = {
      prefix: value.prefix || '',
      namespace: value.namespace || '',
    };
  }

  handlePrefixChange = (e: React.FormEvent<HTMLInputElement>) => {
    const prefix = e.currentTarget.value;
    this.setState({
      prefix,
    });
    this.props.onChange({ ...this.state, prefix });
  };

  handleNamespaceChange = (e: React.FormEvent<HTMLInputElement>) => {
    const namespace = e.currentTarget.value;
    this.setState({
      namespace,
    });
    this.props.onChange({ ...this.state, namespace });
  };

  render() {
    return (
      <Input.Group
        compact
        style={
          {
            width: 'calc(100% - 22px)',
            marginRight: 8,
          } /* icon is 14px + 8px margin-right = 22px*/
        }
      >
        <Input
          style={{ width: '20%' }}
          name={`prefix[${this.props.groupId}]`}
          placeholder="prefix"
          onChange={this.handlePrefixChange}
          value={this.state.prefix}
        />
        <Input
          style={{ width: '80%' }}
          name={`namespace[${this.props.groupId}]`}
          placeholder="namespace"
          onChange={this.handleNamespaceChange}
          value={this.state.namespace}
        />
      </Input.Group>
    );
  }
}

export interface ProjectFormProps {
  form: WrappedFormUtils;
  project?: {
    label: string;
    name: string;
    base?: string;
    prefixMappings?: PrefixMappingGroupInputState[];
  };
  busy?: boolean;
  onSubmit?(project: ProjectFormProps['project']): any;
  mode?: 'create' | 'edit';
}

/**
 * Adaptation of the following example:
 * based on: https://ant.design/components/form/#components-form-demo-dynamic-form-item
 */
const ProjectForm: React.FunctionComponent<ProjectFormProps> = ({
  form,
  project,
  busy = false,
  onSubmit = () => {},
  mode = 'create',
}) => {
  // logic for generating dynamic prefix mapping fields in form
  const currentId =
    project && project.prefixMappings ? project.prefixMappings.length : 0;
  const activeKeys = [...Array(currentId + 1).keys()].slice(1);
  const [prefixMappingKeys, setPrefixMappingKeys] = React.useState({
    currentId,
    activeKeys,
  });

  const { getFieldDecorator, getFieldValue } = form;
  const formItemLayout = {
    labelCol: {
      xs: { span: 24 },
      sm: { span: 5 },
    },
    wrapperCol: {
      xs: { span: 24 },
      sm: { span: 19 },
    },
  };
  const formItemLayoutWithOutLabel = {
    wrapperCol: {
      xs: { span: 24, offset: 0 },
      sm: { span: 19, offset: 5 },
    },
  };

  const add = (k: any) => {
    const { currentId, activeKeys } = prefixMappingKeys;
    const newId: number = currentId + 1;
    // @ts-ignore
    setPrefixMappingKeys({
      currentId: newId,
      activeKeys: [...activeKeys, newId],
    });
  };

  const remove = (k: any) => {
    const { activeKeys } = prefixMappingKeys;
    setPrefixMappingKeys({
      ...prefixMappingKeys,
      activeKeys: activeKeys.filter((key: any) => key !== k),
    });
  };

  const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
    form.validateFields((err, values) => {
      if (!err) {
        onSubmit({
          ...values,
          prefixMappings:
            (values.prefixMappings &&
              values.prefixMappings.filter((p: any) => !!p)) ||
            [],
        });
      }
    });
  };

  const checkPrefix = (
    rule: {},
    value: PrefixMappingGroupInputState,
    callback: (message?: string) => void
  ) => {
    if (value.prefix && value.namespace) {
      callback();
      return;
    }
    callback('You need to specify both prefix and namespace');
  };

  // Dynamic form fields
  const prefixMappingsItems = prefixMappingKeys.activeKeys.map(
    (key: number, index: number) => (
      <Form.Item
        label={index === 0 ? 'Prefix Mappings' : ''}
        {...(index === 0 ? formItemLayout : formItemLayoutWithOutLabel)}
        key={key}
      >
        {getFieldDecorator(`prefixMappings[${key - 1}]`, {
          initialValue: {
            prefix:
              (project &&
                project.prefixMappings &&
                project.prefixMappings[key - 1] &&
                project.prefixMappings[key - 1].prefix) ||
              '',
            namespace:
              (project &&
                project.prefixMappings &&
                project.prefixMappings[key - 1] &&
                project.prefixMappings[key - 1].namespace) ||
              '',
          },
          rules: [{ validator: checkPrefix, required: true }],
        })(
          // @ts-ignore
          <PrefixMappingGroupInput groupId={`${key}`} />
        )}
        {prefixMappingKeys.activeKeys.length > 0 ? (
          <Icon
            className="dynamic-delete-button"
            type="minus-circle-o"
            onClick={() => remove(key)}
          />
        ) : null}
      </Form.Item>
    )
  );
  return (
    <Spin spinning={busy}>
      <Form onSubmit={handleSubmit}>
        <h1>Project: {getFieldValue('name') || (project && project.name)}</h1>
        <Form.Item label="Name" {...formItemLayout}>
          {getFieldDecorator('name', {
            initialValue: project ? project.name : '',
            rules: [{ required: true }],
          })(<Input placeholder="Name" />)}
        </Form.Item>
        <Form.Item label="Label" {...formItemLayout}>
          {getFieldDecorator('label', {
            initialValue: project ? project.label : '',
            rules: [{ required: true }],
          })(<Input placeholder="Label" disabled={mode === 'edit'} />)}
        </Form.Item>
        <Form.Item label="Base" {...formItemLayout}>
          {getFieldDecorator('base', {
            initialValue: project ? project.base : '',
            rules: [{ required: false }],
          })(<Input placeholder="Base" />)}
        </Form.Item>
        {prefixMappingsItems}
        <Form.Item {...formItemLayoutWithOutLabel}>
          <Button type="dashed" onClick={add} style={{ width: '60%' }}>
            <Icon type="plus" /> Add prefix mapping
          </Button>
        </Form.Item>
        <Form.Item {...formItemLayoutWithOutLabel}>
          <Button
            type="primary"
            htmlType="submit"
            className="login-form-button"
          >
            Save
          </Button>
        </Form.Item>
      </Form>
    </Spin>
  );
};

export default Form.create()(ProjectForm);
