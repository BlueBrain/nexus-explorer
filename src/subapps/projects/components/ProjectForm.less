@import '../../../shared/lib.less';

.project-form {
  .ant-form-item-label {
    text-align: left;
  }

  h2 {
    margin-bottom: 1em;
    padding-bottom: 1em;
    border-bottom: 1px solid @border-color-base;
  }

  button {
    margin: 0.5em;
  }
}
