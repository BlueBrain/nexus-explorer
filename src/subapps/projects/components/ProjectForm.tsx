import * as React from 'react';
import { Form, Input, DatePicker, Radio, Row, Col, Button, Spin } from 'antd';
import * as moment from 'moment';

const { Item } = Form;

import './ProjectForm.less';

export type ProjectMetadata = {
  name: string;
  description: string;
  dueDate: string;
  visibility?: string;
  topic?: string;
  hypotheses?: string;
  goals?: string;
  questions?: string;
  type?: string;
};

const ProjectForm: React.FC<{
  onClickCancel(): void;
  onSubmit(data: ProjectMetadata): void;
  busy: boolean;
  project?: ProjectMetadata;
}> = ({ onClickCancel, onSubmit, busy, project }) => {
  const [name, setName] = React.useState<string>(project ? project.name : '');
  const [nameError, setNameError] = React.useState<boolean>(false);
  const [description, setDescription] = React.useState<string>(
    project ? project.description : ''
  );
  const [descriptionError, setDescriptionError] = React.useState<boolean>(
    false
  );
  const [topic, setTopic] = React.useState<string>(
    project && project.topic ? project.topic : ''
  );
  const [hypotheses, setHypotheses] = React.useState<string>(
    project && project.hypotheses ? project.hypotheses : ''
  );
  const [goals, setGoals] = React.useState<string>(
    project && project.goals ? project.goals : ''
  );
  const [questions, setQuestions] = React.useState<string>(
    project && project.questions ? project.questions : ''
  );
  const [dueDate, setDueDate] = React.useState<any>(
    project ? project.dueDate : null
  );
  const [dateError, setDateError] = React.useState<boolean>(false);
  const [visibility, setVisibility] = React.useState(
    project && project.visibility ? project.visibility : 'public'
  );

  const formTitle = project ? 'Edit Project' : 'Create a New Project';
  const formItemLayout = project
    ? {}
    : {
        labelCol: { xs: { span: 24 }, sm: { span: 9 } },
        wrapperCol: { xs: { span: 24 }, sm: { span: 15 } },
      };

  const formColumnLayOut = project
    ? {}
    : {
        xs: 24,
        sm: 24,
        md: 12,
      };

  const isEmptyInput = (value: string) => {
    return value.split(' ').join('') === '';
  };

  const isValidInput = () => {
    let isValid = true;

    if (isEmptyInput(name)) {
      setNameError(true);
      isValid = false;
    } else {
      setNameError(false);
    }

    if (isEmptyInput(description)) {
      setDescriptionError(true);
      isValid = false;
    } else {
      setDescriptionError(false);
    }

    if (!dueDate) {
      setDateError(true);
      isValid = false;
    } else {
      setDateError(false);
    }

    return isValid;
  };

  const onClickSave = () => {
    if (isValidInput()) {
      const data: ProjectMetadata = {
        visibility,
        name,
        description,
        topic,
        hypotheses,
        goals,
        questions,
        dueDate,
        type: 'personal',
      };

      onSubmit(data);
    }
  };

  const onChangeName = (event: any) => {
    setName(event.target.value);
    setNameError(false);
  };

  const onChangeDescription = (event: any) => {
    setDescription(event.target.value);
    setDescriptionError(false);
  };

  const onChangeDate = (date: any) => {
    setDueDate(date);
    setDateError(false);
  };

  return (
    <Form {...formItemLayout} className="project-form">
      <h2>{formTitle}</h2>
      <Spin spinning={busy} tip="Please wait...">
        <Row gutter={24}>
          <Col {...formColumnLayOut}>
            <Item label="Project Type">
              <Radio.Group value="personal">
                <Radio.Button value="personal">Personal</Radio.Button>
                <Radio.Button value="organization" disabled={true}>
                  Organization
                </Radio.Button>
              </Radio.Group>
            </Item>
            <Item
              label="Project Name *"
              validateStatus={nameError ? 'error' : ''}
              help={nameError && 'Please enter a project name'}
            >
              <Input
                value={name}
                onChange={onChangeName}
                disabled={!!project}
              />
            </Item>
            <Item
              label="Project Description *"
              validateStatus={descriptionError ? 'error' : ''}
              help={descriptionError && 'Please enter a description'}
            >
              <Input.TextArea
                value={description}
                onChange={onChangeDescription}
              />
            </Item>
            <Item label="Research Topic">
              <Input
                value={topic}
                onChange={event => setTopic(event.target.value)}
              />
            </Item>
            <Item label="Research Question(s)">
              <Input.TextArea
                value={questions}
                onChange={event => setQuestions(event.target.value)}
              />
            </Item>
          </Col>
          <Col {...formColumnLayOut}>
            <Item label="Project Visibility">
              <Radio.Group
                value={visibility}
                onChange={event => setVisibility(event.target.value)}
              >
                <Radio.Button value="public">Public</Radio.Button>
                <Radio.Button value="private">Private</Radio.Button>
              </Radio.Group>
            </Item>
            <Item
              label="Provisional End Date *"
              validateStatus={dateError ? 'error' : ''}
              help={dateError && 'Please select a due date'}
            >
              <DatePicker value={moment(dueDate)} onChange={onChangeDate} />
            </Item>
            <Item label="Hypotheses">
              <Input.TextArea
                value={hypotheses}
                onChange={event => setHypotheses(event.target.value)}
              />
            </Item>
            <Item label="Goals and deliverables">
              <Input.TextArea
                value={goals}
                onChange={event => setGoals(event.target.value)}
              />
            </Item>
          </Col>
        </Row>
        <Row>
          <Col {...formColumnLayOut} style={{ textAlign: 'left' }}>
            <em>* Mandatory fields</em>
          </Col>
          <Col {...formColumnLayOut} style={{ textAlign: 'right' }}>
            <Button onClick={onClickCancel}>Cancel</Button>
            <Button onClick={onClickSave} type="primary">
              {project ? 'Save' : 'Create'}
            </Button>
          </Col>
        </Row>
      </Spin>
    </Form>
  );
};

export default ProjectForm;
