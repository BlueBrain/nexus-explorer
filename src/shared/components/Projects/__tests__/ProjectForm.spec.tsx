import * as React from 'react';
import { mount } from 'enzyme';
import ProjectForm from '../ProjectForm';

const formComponent = <ProjectForm />;

const wrapper = mount(formComponent);
describe('Project Form component', () => {
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
