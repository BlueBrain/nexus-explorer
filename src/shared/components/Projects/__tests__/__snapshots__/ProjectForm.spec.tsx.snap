// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Form component should render correctly 1`] = `
<Form(ProjectForm)>
  <ProjectForm
    form={
      Object {
        "getFieldDecorator": [Function],
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldProps": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "isSubmitting": [Function],
        "resetFields": [Function],
        "setFields": [Function],
        "setFieldsInitialValue": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
        "validateFieldsAndScroll": [Function],
      }
    }
  >
    <Spin
      size="default"
      spinning={false}
      wrapperClassName=""
    >
      <div
        className="ant-spin-nested-loading"
      >
        <div
          className="ant-spin-container"
          key="container"
        >
          <Form
            hideRequiredMark={false}
            layout="horizontal"
            onSubmit={[Function]}
          >
            <form
              className="ant-form ant-form-horizontal"
              onSubmit={[Function]}
            >
              <FormItem
                colon={true}
                hasFeedback={false}
                label="Label"
                labelCol={
                  Object {
                    "sm": Object {
                      "span": 5,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
                wrapperCol={
                  Object {
                    "sm": Object {
                      "span": 19,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-label"
                      key="label"
                      sm={
                        Object {
                          "span": 5,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-label ant-col-xs-24 ant-col-sm-5"
                      >
                        <label
                          className="ant-form-item-required"
                          htmlFor="label"
                          onClick={[Function]}
                          title="Label"
                        >
                          Label
                        </label>
                      </div>
                    </Col>
                    <Col
                      className="ant-form-item-control-wrapper"
                      key="wrapper"
                      sm={
                        Object {
                          "span": 19,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-19"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Input
                              data-__field={
                                Object {
                                  "name": "label",
                                }
                              }
                              data-__meta={
                                Object {
                                  "initialValue": "",
                                  "name": "label",
                                  "originalProps": Object {
                                    "disabled": false,
                                    "placeholder": "Label",
                                    "type": "text",
                                  },
                                  "ref": null,
                                  "rules": Array [
                                    Object {
                                      "message": "Label must be a phrase without spaces",
                                      "pattern": /\\^\\\\S\\+\\$/g,
                                      "required": true,
                                      "whitespace": true,
                                    },
                                  ],
                                  "trigger": "onChange",
                                  "validate": Array [
                                    Object {
                                      "rules": Array [
                                        Object {
                                          "message": "Label must be a phrase without spaces",
                                          "pattern": /\\^\\\\S\\+\\$/g,
                                          "required": true,
                                          "whitespace": true,
                                        },
                                      ],
                                      "trigger": Array [
                                        "onChange",
                                      ],
                                    },
                                  ],
                                  "valuePropName": "value",
                                }
                              }
                              disabled={false}
                              id="label"
                              onChange={[Function]}
                              placeholder="Label"
                              type="text"
                              value=""
                            >
                              <input
                                className="ant-input"
                                data-__field={
                                  Object {
                                    "name": "label",
                                  }
                                }
                                data-__meta={
                                  Object {
                                    "initialValue": "",
                                    "name": "label",
                                    "originalProps": Object {
                                      "disabled": false,
                                      "placeholder": "Label",
                                      "type": "text",
                                    },
                                    "ref": null,
                                    "rules": Array [
                                      Object {
                                        "message": "Label must be a phrase without spaces",
                                        "pattern": /\\^\\\\S\\+\\$/g,
                                        "required": true,
                                        "whitespace": true,
                                      },
                                    ],
                                    "trigger": "onChange",
                                    "validate": Array [
                                      Object {
                                        "rules": Array [
                                          Object {
                                            "message": "Label must be a phrase without spaces",
                                            "pattern": /\\^\\\\S\\+\\$/g,
                                            "required": true,
                                            "whitespace": true,
                                          },
                                        ],
                                        "trigger": Array [
                                          "onChange",
                                        ],
                                      },
                                    ],
                                    "valuePropName": "value",
                                  }
                                }
                                disabled={false}
                                id="label"
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="Label"
                                type="text"
                                value=""
                              />
                            </Input>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                colon={true}
                hasFeedback={false}
                label="Description"
                labelCol={
                  Object {
                    "sm": Object {
                      "span": 5,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
                wrapperCol={
                  Object {
                    "sm": Object {
                      "span": 19,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-label"
                      key="label"
                      sm={
                        Object {
                          "span": 5,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-label ant-col-xs-24 ant-col-sm-5"
                      >
                        <label
                          className=""
                          htmlFor="description"
                          onClick={[Function]}
                          title="Description"
                        >
                          Description
                        </label>
                      </div>
                    </Col>
                    <Col
                      className="ant-form-item-control-wrapper"
                      key="wrapper"
                      sm={
                        Object {
                          "span": 19,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-19"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Input
                              data-__field={
                                Object {
                                  "name": "description",
                                }
                              }
                              data-__meta={
                                Object {
                                  "initialValue": "",
                                  "name": "description",
                                  "originalProps": Object {
                                    "disabled": false,
                                    "placeholder": "Description",
                                    "type": "text",
                                  },
                                  "ref": null,
                                  "rules": Array [
                                    Object {
                                      "required": false,
                                    },
                                  ],
                                  "trigger": "onChange",
                                  "validate": Array [
                                    Object {
                                      "rules": Array [
                                        Object {
                                          "required": false,
                                        },
                                      ],
                                      "trigger": Array [
                                        "onChange",
                                      ],
                                    },
                                  ],
                                  "valuePropName": "value",
                                }
                              }
                              disabled={false}
                              id="description"
                              onChange={[Function]}
                              placeholder="Description"
                              type="text"
                              value=""
                            >
                              <input
                                className="ant-input"
                                data-__field={
                                  Object {
                                    "name": "description",
                                  }
                                }
                                data-__meta={
                                  Object {
                                    "initialValue": "",
                                    "name": "description",
                                    "originalProps": Object {
                                      "disabled": false,
                                      "placeholder": "Description",
                                      "type": "text",
                                    },
                                    "ref": null,
                                    "rules": Array [
                                      Object {
                                        "required": false,
                                      },
                                    ],
                                    "trigger": "onChange",
                                    "validate": Array [
                                      Object {
                                        "rules": Array [
                                          Object {
                                            "required": false,
                                          },
                                        ],
                                        "trigger": Array [
                                          "onChange",
                                        ],
                                      },
                                    ],
                                    "valuePropName": "value",
                                  }
                                }
                                disabled={false}
                                id="description"
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="Description"
                                type="text"
                                value=""
                              />
                            </Input>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                colon={true}
                hasFeedback={false}
                label="Base"
                labelCol={
                  Object {
                    "sm": Object {
                      "span": 5,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
                wrapperCol={
                  Object {
                    "sm": Object {
                      "span": 19,
                    },
                    "xs": Object {
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-label"
                      key="label"
                      sm={
                        Object {
                          "span": 5,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-label ant-col-xs-24 ant-col-sm-5"
                      >
                        <label
                          className=""
                          htmlFor="base"
                          onClick={[Function]}
                          title="Base"
                        >
                          Base
                        </label>
                      </div>
                    </Col>
                    <Col
                      className="ant-form-item-control-wrapper"
                      key="wrapper"
                      sm={
                        Object {
                          "span": 19,
                        }
                      }
                      xs={
                        Object {
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-19"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Input
                              data-__field={
                                Object {
                                  "name": "base",
                                }
                              }
                              data-__meta={
                                Object {
                                  "initialValue": "",
                                  "name": "base",
                                  "originalProps": Object {
                                    "disabled": false,
                                    "placeholder": "Base",
                                    "type": "text",
                                  },
                                  "ref": null,
                                  "rules": Array [
                                    Object {
                                      "required": false,
                                    },
                                  ],
                                  "trigger": "onChange",
                                  "validate": Array [
                                    Object {
                                      "rules": Array [
                                        Object {
                                          "required": false,
                                        },
                                      ],
                                      "trigger": Array [
                                        "onChange",
                                      ],
                                    },
                                  ],
                                  "valuePropName": "value",
                                }
                              }
                              disabled={false}
                              id="base"
                              onChange={[Function]}
                              placeholder="Base"
                              type="text"
                              value=""
                            >
                              <input
                                className="ant-input"
                                data-__field={
                                  Object {
                                    "name": "base",
                                  }
                                }
                                data-__meta={
                                  Object {
                                    "initialValue": "",
                                    "name": "base",
                                    "originalProps": Object {
                                      "disabled": false,
                                      "placeholder": "Base",
                                      "type": "text",
                                    },
                                    "ref": null,
                                    "rules": Array [
                                      Object {
                                        "required": false,
                                      },
                                    ],
                                    "trigger": "onChange",
                                    "validate": Array [
                                      Object {
                                        "rules": Array [
                                          Object {
                                            "required": false,
                                          },
                                        ],
                                        "trigger": Array [
                                          "onChange",
                                        ],
                                      },
                                    ],
                                    "valuePropName": "value",
                                  }
                                }
                                disabled={false}
                                id="base"
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="Base"
                                type="text"
                                value=""
                              />
                            </Input>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                colon={true}
                hasFeedback={false}
                wrapperCol={
                  Object {
                    "sm": Object {
                      "offset": 5,
                      "span": 19,
                    },
                    "xs": Object {
                      "offset": 0,
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-control-wrapper"
                      key="wrapper"
                      sm={
                        Object {
                          "offset": 5,
                          "span": 19,
                        }
                      }
                      xs={
                        Object {
                          "offset": 0,
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-19 ant-col-sm-offset-5"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Button
                              block={false}
                              ghost={false}
                              loading={false}
                              onClick={[Function]}
                              style={
                                Object {
                                  "width": "60%",
                                }
                              }
                              type="dashed"
                            >
                              <Wave>
                                <button
                                  className="ant-btn ant-btn-dashed"
                                  onClick={[Function]}
                                  style={
                                    Object {
                                      "width": "60%",
                                    }
                                  }
                                  type="button"
                                >
                                  <Icon
                                    key=".0"
                                    type="plus"
                                  >
                                    <i
                                      className="anticon anticon-plus"
                                    >
                                      <IconReact
                                        className=""
                                        type="plus-o"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          className=""
                                          data-icon="plus"
                                          fill="currentColor"
                                          height="1em"
                                          key="svg-plus"
                                          viewBox="64 64 896 896"
                                          width="1em"
                                        >
                                          <path
                                            d="M848 474H550V152h-76v322H176c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h298v322h76V550h298c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
                                            key="svg-plus-svg-0"
                                          />
                                        </svg>
                                      </IconReact>
                                    </i>
                                  </Icon>
                                  <span
                                    key=".1"
                                  >
                                     Add API mapping
                                  </span>
                                </button>
                              </Wave>
                            </Button>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
              <FormItem
                colon={true}
                hasFeedback={false}
                wrapperCol={
                  Object {
                    "sm": Object {
                      "offset": 5,
                      "span": 19,
                    },
                    "xs": Object {
                      "offset": 0,
                      "span": 24,
                    },
                  }
                }
              >
                <Row
                  className="ant-form-item"
                  gutter={0}
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <Col
                      className="ant-form-item-control-wrapper"
                      key="wrapper"
                      sm={
                        Object {
                          "offset": 5,
                          "span": 19,
                        }
                      }
                      xs={
                        Object {
                          "offset": 0,
                          "span": 24,
                        }
                      }
                    >
                      <div
                        className="ant-form-item-control-wrapper ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-19 ant-col-sm-offset-5"
                      >
                        <div
                          className="ant-form-item-control"
                        >
                          <span
                            className="ant-form-item-children"
                          >
                            <Button
                              block={false}
                              className="login-form-button"
                              ghost={false}
                              htmlType="submit"
                              loading={false}
                              type="primary"
                            >
                              <Wave>
                                <button
                                  className="ant-btn login-form-button ant-btn-primary"
                                  onClick={[Function]}
                                  type="submit"
                                >
                                  <span
                                    key=".0"
                                  >
                                    Save
                                  </span>
                                </button>
                              </Wave>
                            </Button>
                          </span>
                          <Animate
                            animation={Object {}}
                            component=""
                            componentProps={Object {}}
                            key="help"
                            onAppear={[Function]}
                            onEnd={[Function]}
                            onEnter={[Function]}
                            onLeave={[Function]}
                            transitionAppear={true}
                            transitionEnter={true}
                            transitionLeave={true}
                            transitionName="show-help"
                          />
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
              </FormItem>
            </form>
          </Form>
        </div>
      </div>
    </Spin>
  </ProjectForm>
</Form(ProjectForm)>
`;
